### PR TITLE
Remove JuMP dependency

### PR DESCRIPTION
**Pull request by blegat**
This is an attempt to simplify the code and remove the JuMP dependency. The hope is to simply use a `cache` and `copy_to` to generalize Juniper to any constraint type.

### CODE REVIEW

**Comment by ccoffrin** [comment]
Nice, I did not realize this was even possible!

**Comment by blegat** [comment]
Another advantage for `copy_to` is that it will be faster than the current one: https://github.com/lanl-ansi/Juniper.jl/blob/eca68b3425a9b7b5ebcd165e070375c824a38997/src/model.jl#L43 which has type unstability issues

**Comment by blegat** [comment]
Should be ready for review :)
With this PR, Juniper should now work for any constraint type (so it should for instance work for MISDP if the nlp solver supports PSD constraints) although it probably still needs some work to support other mixed-integer sets such as `Indicator`, `SOS`, Semiinteger`, ... as currently, it will try to copy it to the nonlinear solver and it will result in an `UnsupportedConstraint` error. We probably just need to filter out but I'll leave that for another PR.

File: src/bb_inits_and_defaults.jl [hunk 1]
```diff
@@ -196,10 +192,6 @@ function init_juniper_problem!(jp::JuniperProblem, model::MOI.AbstractOptimizer)
             int_i += 1
         end
     end
-    jp.num_l_constr = num_linear_le_constraints+num_linear_ge_constraints+num_linear_eq_constraints
```
Comment (Wikunia) [inline src/bb_inits_and_defaults.jl:199]: is this information lost with the new change?

**Review by Wikunia — COMMENTED** [review]
Thanks a lot! Overall it looks much cleaner. 

Just some basic questions for now beside the one comment I've added:
Will MOI use bridges automatically for the MIP part in the feasibility pump if for example the solver doesn't support `>=` just `<=` ? That's more a general MOI question as you know that better than me :wink:

It seems like we don't need the extra register for user defined functions anymore, right?
If that is actually the case first of all I'm amazed :tada: 
Would be great if you can remove https://lanl-ansi.github.io/Juniper.jl/stable/options/#registered_functions::Union%7BNothing,Vector%7BRegisteredFunction%7D%7D-%5Bnothing%5D from the docs then as well.

@ccoffrin before we merge it would be in general nice to run the regression tests again to see if any problem fails or if there is speedup/slowdown.

File: src/bb_inits_and_defaults.jl (hunk 1 above)
Comment (blegat) [reply src/bb_inits_and_defaults.jl:199]: Yes, it's not so well defined for a generic MOI model. For instance, do you count NormOneConstraint as linear ? As how many ? If you have a constraint in `MOI.Zeros(1000)` is it one constraint or `1000` ?
As it's only used in printing, we could add this back later with a more general printing that says, e.g. how many constraint there is of each type.

**Comment by blegat** [comment]
> Will MOI use bridges automatically for the MIP part in the feasibility pump if for example the solver doesn't support >= just <= ?

Yes, we use `MOI.instantiate(solver, with_bridge_type=Float64)` which adds a bridge layer will all the bridges just  like you have with JuMP.

> It seems like we don't need the extra register for user defined functions anymore, right?

Yes :) will remove

**Comment by ccoffrin** [comment]
@Wikunia, sure thing I'll do the full regression.

**Comment by ccoffrin** [comment]
While testing this I was wondering if JuMP can now be moved from deps to extras on the Project.toml?

**Comment by blegat** [comment]
> I was wondering if JuMP can now be moved from deps to extras on the Project.toml?

Yes, done :)

**Comment by Wikunia** [comment]
I have some great news @blegat & @ccoffrin 

With one processor the MOI version is quite a bit better
juniper_moi

With four cores it is better earlier on and then they solve the same number of instances
juniper_moi-p4

For 8 cores it looks quite similar to the four cores one ;)

**Comment by ccoffrin** [comment]
Nice!  Shall we merge and release then?

Maybe we should tag the current master before merging?

**Comment by Wikunia** [comment]
You mean current master as v0.6.6 and this as v0.7.0 ?

**Comment by ccoffrin** [comment]
You are correct. Nothing new on master, no v0.6.6 release is required.

**Pull request merged**